File: Count the Number of Complete Components.py
# https://leetcode.com/problems/count-the-number-of-complete-components/?envType=daily-question&envId=2025-03-22
# Count the Number of Complete Components

class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        
        graph = [set() for _ in range(n)]
        for u, v in edges:
            graph[u].add(v)
            graph[v].add(u)
        
        visited = [False] * n
        complete_count = 0
        
        def bfs(start: int) -> List[int]:
            queue = deque([start])
            comp = []
            visited[start] = True
            while queue:
                node = queue.popleft()
                comp.append(node)
                for neighbor in graph[node]:
                    if not visited[neighbor]:
                        visited[neighbor] = True
                        queue.append(neighbor)
            return comp
        
        for i in range(n):
            if not visited[i]:
                comp = bfs(i)
                comp_set = set(comp)
                k = len(comp)
                is_complete = True
                for node in comp:
                    if len(graph[node].intersection(comp_set)) != k - 1:
                        is_complete = False
                        break
                if is_complete:
                    complete_count += 1
        
        return complete_count

File: Count the Number of Complete Components.js
// https://leetcode.com/problems/count-the-number-of-complete-components/?envType=daily-question&envId=2026-07-11
// Count the Number of Complete Components

/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number}
 */
var countCompleteComponents = function(n, edges) {
    
    const graph = Array.from({ length: n }, () => new Set());

    for (const [u, v] of edges) {
        graph[u].add(v);
        graph[v].add(u);
    }

    const visited = new Array(n).fill(false);
    let completeCount = 0;

    function bfs(start) {
        const queue = [start];
        let head = 0;
        const component = [];

        visited[start] = true;

        while (head < queue.length) {
            const node = queue[head++];
            component.push(node);

            for (const neighbor of graph[node]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.push(neighbor);
                }
            }
        }

        return component;
    }

    for (let i = 0; i < n; i++) {
        if (!visited[i]) {
            const component = bfs(i);
            const componentSet = new Set(component);
            const k = component.length;

            let isComplete = true;

            for (const node of component) {
                let neighborCount = 0;

                for (const neighbor of graph[node]) {
                    if (componentSet.has(neighbor)) {
                        neighborCount++;
                    }
                }

                if (neighborCount !== k - 1) {
                    isComplete = false;
                    break;
                }
            }

            if (isComplete) {
                completeCount++;
            }
        }
    }

    return completeCount;
};